lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+
+        loop_state = (
+            loop_state_main,
+            loop_state_r_est,
+            loop_state_anorm,
+            loop_state_stopping,
+            loop_state_vecs,
+        )
+
+        def condfun(loop_state):
+            (
+                loop_state_main,
+                loop_state_r_est,
+                loop_state_anorm,
+                loop_state_stopping,
+                loop_state_vecs,
+            ) = loop_state
+            (
+                itn,
+                alpha,
+                beta,
+                zetabar,
+                alphabar,
+                rho,
+                rhobar,
+                cbar,
+                sbar,
+            ) = loop_state_main
+            (
+                betadd,
+                betad,
+                rhodold,
+                tautildeold,
+                thetatilde,
+                zeta,
+                d,
+            ) = loop_state_r_est
+            (normA2, maxrbar, minrbar, normA, condA) = loop_state_anorm
+            (istop, normr, normb) = loop_state_stopping
+            (x, u, v, h, hbar) = loop_state_vecs
+
+            # Compute norms for convergence testing.
+            normar = abs(zetabar)
+            normx = self.norm(x)
+
+            # Now use these norms to estimate certain other quantities,
+            # some of which will be small near a solution.
+            test1 = normr / normb
+            test2 = lax.select((normA * normr) != 0, normar / (normA * normr), jnp.inf)
+            test3 = 1 / condA
+            t1 = test1 / (1 + normA * normx / normb)
+            rtol = self.btol + self.atol * normA * normx / normb
+
+            # bitmask for istop for possibly multiple conditions?
+
+            # x is an approximate solution to A@x = b, according to atol and btol.
+            istop += (test1 <= rtol).astype(int) * 2**1
+            # x approximately solves the least-squares problem according to atol.
+            istop += (test2 <= self.atol).astype(int) * 2**2
+            # cond(A) seems to be greater than conlim
+            istop += (test3 <= ctol).astype(int) * 2**3
+
+            # The following tests guard against extremely small values of atol, btol
+            # or ctol.  (The user may have set any or all of the parameters atol, btol,
+            # conlim  to 0.) The effect is equivalent to the normal tests using
+            # atol = eps,  btol = eps,  conlim = 1/eps.
+
+            # x is an approximate solution to A@x = b, according to atol=btol=eps.
+            istop += (1 + t1 <= 1).astype(int) * 2**4
+            # x approximately solves the least-squares problem according to atol=eps.
+            istop += (1 + test2 <= 1).astype(int) * 2**5
+            # cond(A) seems to be greater than 1/eps
+            istop += (1 + test3 <= 1).astype(int) * 2**6
+
+            # maxiter exceeded
+            istop += (itn >= max_steps).astype(int) * 2**7
+
+            return istop == 0
```

Comment:
This seems like an odd way to test multiple conditions. I'm guessing this is coming from some older system which didn't support bool types.

I think here we probably just want `return (itn < max_steps) & (1 + test3 > 1) & ...` ?

Reply:
The main reason for this was so that after exiting the loop we actually know which condition caused the exit, since there are in theory a few different options. Another option would be to add specific flags to the state that get set but the state is already pretty huge...